ts/nni_manager/common/command_channel/http.ts

```diff
@@ -0,0 +1,140 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+/**
+ *  HTTP based command channel. Inefficient but simple.
+ *
+ *  Client to server command is implemented by naive PUT request.
+ *
+ *  Server to client command is implemented by polling with GET requests.
+ *  The client should repeat GET requests without intervals when it is receiving commands.
+ *  If the server has an outgoing command, it will be sent as response.
+ *  If there is no command, the server will wait for 1 second and response 408 "Request Timeout" status code.
+ **/
+
+import { EventEmitter } from 'events';
+
+import { Request, Response } from 'express';
+
+import { Deferred } from 'common/deferred';
+import globals from 'common/globals';
+import { Logger, getLogger } from 'common/log';
+import type { Command, CommandChannelServer } from './interface';
+
+let timeoutMilliseconds = 1000;
+
+const HttpRequestTimeout = 408;
+const HttpGone = 410;
+
+export class HttpChannelServer implements CommandChannelServer {
+    private emitter: EventEmitter = new EventEmitter();
+    private log: Logger;
+    // the server can only send commands when the client requests, so it needs a queue
+    private outgoingQueues: Map<string, CommandQueue> = new Map();
+    private path: string;
+    private serving: boolean = false;
+
+    constructor(name: string, urlPath: string) {
+        this.log = getLogger(`HttpChannelManager.${name}`);
+        this.path = urlPath;
+    }
+
+    public start(): void {
+        this.serving = true;
+        const channelPath = globals.rest.urlJoin(this.path, ':channel');
+        globals.rest.registerSyncHandler('GET', channelPath, this.handleGet.bind(this));
+        globals.rest.registerSyncHandler('PUT', channelPath, this.handlePut.bind(this));
+    }
+
+    public shutdown(): void {
+        this.serving = false;
+        this.outgoingQueues.forEach(queue => { queue.clear(); });
```

Comment:
It looks not symmetric to `start()`. For example, `start()` has registerHandler, but shudown doesn't have unregister.

Reply:
That's why it's called shutdown rather than stop.
Removing handler is not an easy task for express. I have not seen needs for "stop" so far so it's not worthy the efforts.